peregrine/resources/submission/graphql/node.py

```diff
@@ -795,6 +797,67 @@ def _make_inner_meta_type():
     return gql_object
 
 
+class NodeCounter:
+    """A workaround for performance issue in PXD-2336.
+
+    With large dictionary of hundreds of nodes, constructing count queries can be a time
+    consuming task with SQLAlchemy. This workaround uses raw SQL and combines all simple
+    count queries into one to reduce round-trip time to database.
+    """
+    query_tpl = "(SELECT count(*) FROM %s WHERE _props->>'project_id' = :%s) AS %s"
+
+    def __init__(self):
+        self._project_ids = {}  # map of project_ids to their placeholder names
+        self._queries = []  # list of (key, promise, scalar subquery)
+
+    @classmethod
+    def current(cls):
+        if not hasattr(fg, 'node_counter'):
+            fg.node_counter = cls()
+        return fg.node_counter
+
+    def add_count(self, cls, args):
+        # escape non-trivial cases
+        if cls != psqlgraph.Node and not hasattr(cls, 'project_id') or \
```

Comment:
this conditional is kind of hard to read, could you split up some of these checks to make it more readable?

Reply:
sure thing!